### PR TITLE
1459 introducing macro parameters

### LINKED ISSUE #1459: create macroParameters in par file to reduce par size

### PR DESCRIPTION
**Pull request by marcochiaramello**
Closes #1459

### CODE REVIEW

File: dynawo/sources/API/PAR/PARMacroParameterSet.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+  /**
+   * @brief getter of map of parameters
+   * @returns reference to map of parameters
+   */
+  inline std::map<std::string, boost::shared_ptr<Parameter> >& getParameters() {return parameters_;}
+
+ private:
+  std::string id_;                                                   ///< id of the macroParameterSet
+  std::map<std::string, boost::shared_ptr<Reference> > references_;  ///< map of references (key->name, value->reference)
```
Comment (rosiereflo) [inline dynawo/sources/API/PAR/PARMacroParameterSet.h:77]: why using maps and not hashmaps?

File: dynawo/sources/API/PAR/PARParametersSet.h [hunk 2]
```diff
@@ -191,6 +192,14 @@ class ParametersSet : public boost::enable_shared_from_this<ParametersSet> {
    */
   boost::shared_ptr<ParametersSet> addParameter(const boost::shared_ptr<Parameter>& param);
 
+  /**
+   * @brief Add a new macroParSet in macroParSets_ map if it does not already exists
```
Comment (rosiereflo) [inline dynawo/sources/API/PAR/PARParametersSet.h:196]: Please avoid refering to internal structures in description: Add a new macroParSet to the parameter set

Comment (rosiereflo) [reply dynawo/sources/API/PAR/PARParametersSet.h:196]: Please also mentions that this throws if the id is already used

File: dynawo/sources/API/PAR/PARParametersSet.h [hunk 3]
```diff
@@ -498,6 +529,7 @@ class ParametersSet : public boost::enable_shared_from_this<ParametersSet> {
   std::string filepath_;                                                        /**< Parameters' set filepath */
   std::map<std::string, boost::shared_ptr<Parameter> > parameters_;             /**< Map of the parameters */
   boost::unordered_map<std::string, boost::shared_ptr<Reference> > references_; /**< Map of the references */
+  std::map<std::string, boost::shared_ptr<MacroParSet> > macroParSets_;         ///< Map of the macroParSet
```
Comment (rosiereflo) [inline dynawo/sources/API/PAR/PARParametersSet.h:532]: why using a map and not a hashmap?

File: dynawo/sources/API/PAR/PARParametersSetCollection.cpp [hunk 4]
```diff
@@ -47,11 +48,45 @@ ParametersSetCollection::addParametersSet(shared_ptr<ParametersSet> paramSet, bo
   parametersSets_[id] = paramSet;
 }
 
+void
+ParametersSetCollection::addMacroParameterSet(shared_ptr<MacroParameterSet> macroParamSet, bool force) {
+  assert(macroParamSet && "impossible to add null macro parameter set pointer to collection");
+  string id = macroParamSet->getId();
+  if (hasMacroParametersSet(id)) {
+    if (force) {
+      // @todo : force / Create custom id, change parameters set one and adds it to the map
```
Comment (rosiereflo) [inline dynawo/sources/API/PAR/PARParametersSetCollection.cpp:57]: Please solve this todo or remove the force option

File: dynawo/sources/API/PAR/PARParametersSetCollection.cpp [hunk 5]
```diff
@@ -47,11 +48,45 @@ ParametersSetCollection::addParametersSet(shared_ptr<ParametersSet> paramSet, bo
...
+}
+
 shared_ptr<ParametersSet>
 ParametersSetCollection::getParametersSet(const string& id) {
   map< string, shared_ptr<ParametersSet> >::iterator itParamSet = parametersSets_.find(id);
   if (itParamSet == parametersSets_.end())
     throw DYNError(DYN::Error::API, ParametersSetNotFound, id);
+  // if a macroParSet is defined in the set, we add the references and parameters associated
+  if (itParamSet->second->hasMacroParSet()) {
+    map< string, shared_ptr<MacroParSet> >::iterator itMacroParSet;
```
Comment (rosiereflo) [inline dynawo/sources/API/PAR/PARParametersSetCollection.cpp:72]: iterator should always be defined within the for loop and not before

File: dynawo/sources/API/PAR/PARParametersSetCollection.cpp [hunk 6]
```diff
@@ -47,11 +48,45 @@ ParametersSetCollection::addParametersSet(shared_ptr<ParametersSet> paramSet, bo
...
+    for (itMacroParSet = itParamSet->second->getMacroParSets().begin(); itMacroParSet != itParamSet->second->getMacroParSets().end(); itMacroParSet++) {
+      if (hasMacroParametersSet(itMacroParSet->first)) {
+        std::map<std::string, boost::shared_ptr<Reference> >::iterator itReference;
```
Comment (rosiereflo) [inline dynawo/sources/API/PAR/PARParametersSetCollection.cpp:75]: same as above

File: dynawo/sources/API/PAR/PARParametersSetCollection.cpp [hunk 7]
```diff
@@ -47,11 +48,45 @@ ParametersSetCollection::addParametersSet(shared_ptr<ParametersSet> paramSet, bo
...
+
 shared_ptr<ParametersSet>
 ParametersSetCollection::getParametersSet(const string& id) {
   map< string, shared_ptr<ParametersSet> >::iterator itParamSet = parametersSets_.find(id);
   if (itParamSet == parametersSets_.end())
     throw DYNError(DYN::Error::API, ParametersSetNotFound, id);
+  // if a macroParSet is defined in the set, we add the references and parameters associated
+  if (itParamSet->second->hasMacroParSet()) {
+    map< string, shared_ptr<MacroParSet> >::iterator itMacroParSet;
+    for (itMacroParSet = itParamSet->second->getMacroParSets().begin(); itMacroParSet != itParamSet->second->getMacroParSets().end(); itMacroParSet++) {
```
Comment (rosiereflo) [inline dynawo/sources/API/PAR/PARParametersSetCollection.cpp:73]: I generally write my for loops as follows for performance reasons:
for (map< string, shared_ptr<MacroParSet> >::iterator itMacroParSet = itParamSet->second->getMacroParSets().begin(), **itEnd = itParamSet->second->getMacroParSets().end()**; itMacroParSet != **itEnd**; **++i**tMacroParSet)

File: dynawo/sources/API/PAR/PARParametersSetCollection.h [hunk 8]
```diff
@@ -66,13 +67,38 @@ class ParametersSetCollection {
    */
   bool hasParametersSet(const std::string& id);
 
+  /**
+   * @brief Check if a macroParameterSet set is in the collection
+   *
+   * @param[in] id Name of the macroParameterSet
+   * @returns Existence of macroParameterSet in the collection
+   */
+  bool hasMacroParametersSet(const std::string& id);
```
Comment (rosiereflo) [inline dynawo/sources/API/PAR/PARParametersSetCollection.h:76]: why not const?

File: dynawo/sources/API/PAR/PARParametersSetCollection.h [hunk 9]
```diff
@@ -172,6 +198,7 @@ class ParametersSetCollection {
 
  private:
   std::map<std::string, boost::shared_ptr<ParametersSet> > parametersSets_; /**< Map of the parameters set */
+  std::map<std::string, boost::shared_ptr<MacroParameterSet> > macroParametersSets_;  ///< Map of macroParametersSet (key->id, value->MacroParameterSet)
```
Comment (rosiereflo) [inline dynawo/sources/API/PAR/PARParametersSetCollection.h:201]: always the same question about maps

File: dynawo/sources/API/PAR/PARMacroParameterSet.h [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+   * @brief parameter adder
+   * @param[in] parameter to be added
+   */
+  void addParameter(boost::shared_ptr<Parameter> parameter);
+
+  /**
+   * @brief getter of map of references
+   * @returns reference to map of references
+   */
+  inline std::map<std::string, boost::shared_ptr<Reference> >& getReferences() {return references_;}
```
Comment (rosiereflo) [inline dynawo/sources/API/PAR/PARMacroParameterSet.h:67]: It would be cleaner to add a begin, end and whatever operations are done  on this map rather than exposing its type and letting anyone edit it without the class control (same for all similar gets)

File: dynawo/sources/API/PAR/xsd/parameters.xsd [hunk 11]
```diff
@@ -19,14 +19,15 @@
 
   <xs:element name="parametersSet">
     <xs:complexType>
-      <xs:sequence>
+      <xs:choice minOccurs="0" maxOccurs="unbounded">
```
Comment (rosiereflo) [inline dynawo/sources/API/PAR/xsd/parameters.xsd:22]: a choice means that only one of the elements is given, you have to use sequence here if I'm not mistaken

File: dynawo/sources/Common/DYNTimer.h [hunk 12]
```diff
@@ -27,7 +27,7 @@
 #include <boost/timer.hpp>
 #include <boost/core/noncopyable.hpp>
 
-// #define PRINT_TIMERS
+#define PRINT_TIMERS
```
Comment (rosiereflo) [inline dynawo/sources/Common/DYNTimer.h:30]: please revert this

File: dynawo/sources/Common/Dictionaries/DYNError_en_GB.dic [hunk 13]
```diff
@@ -260,12 +260,16 @@ ExternalVariableIDNotUnique =             external variable id (%1%) is not uniq
 //---JOB---------------------
 CriteriaStepError           =             criteria step should be a positive integer (value found: %1%)
 //---PAR---------------------
+MacroParameterSetAlreadyExists =          the macroParameterSet %1% already exists in collection
```
Comment (rosiereflo) [inline dynawo/sources/Common/Dictionaries/DYNError_en_GB.dic:263]: you will need to update accordingly the french dictionary also

File: dynawo/sources/Modeler/Common/DYNDynamicData.cpp [hunk 14]
```diff
@@ -251,6 +252,9 @@ DynamicData::associateParameters() {
 
 shared_ptr<ParametersSet>
 DynamicData::getParametersSet(const string& modelId, const string& parFile, const string& parId) {
+  #if defined(_DEBUG_) || defined(PRINT_TIMERS)
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/Common/DYNDynamicData.cpp:255]: not needed

File: dynawo/sources/Modeler/Common/DYNSubModel.cpp [hunk 15]
```diff
@@ -484,6 +484,7 @@ SubModel::setParameterFromSet(ParameterModeler& parameter, const shared_ptr<para
                               const parameterOrigin_t& origin) {
   if (parametersSet) {
     const string& parName = parameter.getName();
+    // std::cout << parName << std::endl;
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/Common/DYNSubModel.cpp:487]: not needed

File: dynawo/sources/Modeler/Common/DYNSubModel.h [hunk 16]
```diff
@@ -1437,6 +1437,7 @@ class SubModel {
    * @param parameter: parameter to be set
    */
   inline void setParameterFromPARFile(ParameterModeler& parameter) {
+    // std::cout << parameter.getName() << std::endl;
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/Common/DYNSubModel.h:1440]: not needed

File: dynawo/sources/Models/CPP/Components/ModelLoadRestorativeWithLimits/DYNModelLoadRestorativeWithLimits.cpp [hunk 17]
```diff
@@ -135,7 +135,7 @@ namespace DYN {
     variables.push_back(VariableNativeFactory::createState("Ii_value", FLOW));
     variables.push_back(VariableNativeFactory::createState("PPu_value", CONTINUOUS));
     variables.push_back(VariableNativeFactory::createState("QPu_value", CONTINUOUS));
-    variables.push_back(VariableNativeFactory::createCalculated("state_value", CONTINUOUS));
+    variables.push_back(VariableNativeFactory::createCalculated("state_value", INTEGER));
```
Comment (rosiereflo) [inline dynawo/sources/Models/CPP/Components/ModelLoadRestorativeWithLimits/DYNModelLoadRestorativeWithLimits.cpp:138]: shouldn't this be done in another PR as unrelated?

**Review by rosiereflo — CHANGES_REQUESTED** [review]
Unit tests + tnr showing the syntax are missing

File: dynawo/sources/Modeler/Common/DYNSubModel.cpp (hunk 15 above)
Comment (marcochiaramello) [reply dynawo/sources/Modeler/Common/DYNSubModel.cpp:487]: yes, sorry, wrong manipulation with git, removed

File: dynawo/sources/Modeler/Common/DYNSubModel.h (hunk 16 above)
Comment (marcochiaramello) [reply dynawo/sources/Modeler/Common/DYNSubModel.h:1440]: solved

File: dynawo/sources/Models/CPP/Components/ModelLoadRestorativeWithLimits/DYNModelLoadRestorativeWithLimits.cpp (hunk 17 above)
Comment (marcochiaramello) [reply dynawo/sources/Models/CPP/Components/ModelLoadRestorativeWithLimits/DYNModelLoadRestorativeWithLimits.cpp:138]: wrong manipulation with git, sorry, solved

File: dynawo/sources/Common/DYNTimer.h (hunk 12 above)
Comment (marcochiaramello) [reply dynawo/sources/Common/DYNTimer.h:30]: done

File: dynawo/sources/API/PAR/PARMacroParameterSet.h (hunk 1 above)
Comment (marcochiaramello) [reply dynawo/sources/API/PAR/PARMacroParameterSet.h:77]: all the map used are dumped in the xml at export level, so that is better to keep them ordered

File: dynawo/sources/API/PAR/PARParametersSet.h (hunk 3 above)
Comment (marcochiaramello) [reply dynawo/sources/API/PAR/PARParametersSet.h:532]: same as above

File: dynawo/sources/Modeler/Common/DYNDynamicData.cpp (hunk 14 above)
Comment (marcochiaramello) [reply dynawo/sources/Modeler/Common/DYNDynamicData.cpp:255]: ok, done

File: dynawo/sources/API/PAR/PARParametersSet.h (hunk 2 above)
Comment (marcochiaramello) [reply dynawo/sources/API/PAR/PARParametersSet.h:196]: ok, thanks

**Comment by marcochiaramello** [comment]
unit test coverage for PAR_unittest-tests -> 95.2 %

File: dynawo/sources/API/PAR/PARMacroParSetFactory.cpp [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+
+using std::string;
+
+using boost::shared_ptr;
+
+namespace parameters {
+
+shared_ptr<MacroParSet>
+MacroParSetFactory::newMacroParSet(const string& id) {
+  return shared_ptr<MacroParSet>(new MacroParSet(id));
```
Comment (rosiereflo) [inline dynawo/sources/API/PAR/PARMacroParSetFactory.cpp:32]: factory are generally used to hide an Impl class from the external world. As it is not the case here I am not sure if this is useful.

File: dynawo/sources/API/PAR/PARParametersSetCollection.cpp [hunk 19]
```diff
@@ -47,11 +48,40 @@ ParametersSetCollection::addParametersSet(shared_ptr<ParametersSet> paramSet, bo
...
+  macroParametersSets_[id] = macroParamSet;
+}
+
 shared_ptr<ParametersSet>
 ParametersSetCollection::getParametersSet(const string& id) {
   map< string, shared_ptr<ParametersSet> >::iterator itParamSet = parametersSets_.find(id);
   if (itParamSet == parametersSets_.end())
     throw DYNError(DYN::Error::API, ParametersSetNotFound, id);
+  // if a macroParSet is defined in the set, we add the references and parameters associated
+  if (itParamSet->second->hasMacroParSet()) {
```
Comment (rosiereflo) [inline dynawo/sources/API/PAR/PARParametersSetCollection.cpp:67]: This will be done everythime we read the parameter even if its used by many models. Shouldn't we factorize this and do it only once?

File: dynawo/sources/Modeler/Common/DYNDynamicData.cpp [hunk 20]
```diff
@@ -57,6 +57,7 @@
 #include "DYDIterators.h"
 #include "DYDMacroConnect.h"
 #include "DYDMacroConnector.h"
+#include "DYNTimer.h"
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/Common/DYNDynamicData.cpp:60]: useless

File: nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_MacroParameterSet/IEEE14.jobs [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+    This Source Code Form is subject to the terms of the Mozilla Public
+    License, v. 2.0. If a copy of the MPL was not distributed with this
+    file, you can obtain one at http://mozilla.org/MPL/2.0/.
+    SPDX-License-Identifier: MPL-2.0
+
+    This file is part of Dynawo, an hybrid C++/Modelica open source time domain
+    simulation tool for power systems.
+-->
+<dyn:jobs xmlns:dyn="http://www.rte-france.com/dynawo">
+  <dyn:job name="IEEE14 - MacroStaticRefs">
```
Comment (rosiereflo) [inline nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_MacroParameterSet/IEEE14.jobs:15]: Please adapt the name of the job

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/API/PAR/PARMacroParSetFactory.cpp (hunk 18 above)
Comment (marcochiaramello) [reply dynawo/sources/API/PAR/PARMacroParSetFactory.cpp:32]: to be honest, I had the same thought, but then I have observed that was done like this also for macroConnectors, macroStaticReference, etc.. if you wish, we can get rid of this (also for dyd objects?)
@adriguir @gautierbureau your opinions on this?

File: nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_MacroParameterSet/IEEE14.jobs (hunk 21 above)
Comment (marcochiaramello) [reply nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_MacroParameterSet/IEEE14.jobs:15]: ok, thanks!

File: dynawo/sources/Modeler/Common/DYNDynamicData.cpp (hunk 20 above)
Comment (marcochiaramello) [reply dynawo/sources/Modeler/Common/DYNDynamicData.cpp:60]: done, thanks

File: dynawo/sources/API/PAR/PARParametersSetCollection.cpp (hunk 19 above)
Comment (marcochiaramello) [reply dynawo/sources/API/PAR/PARParametersSetCollection.cpp:75]: solved separating it in a different method called only once for each par file

**Review by rosiereflo — APPROVED** [review]

File: dynawo/sources/API/PAR/PARMacroParSetFactory.cpp (hunk 18 above)
Comment (marcochiaramello) [reply dynawo/sources/API/PAR/PARMacroParSetFactory.cpp]: after short discussion, factories have been removed for this commit

**Review by rosiereflo — APPROVED** [review]

**Review by gautierbureau — APPROVED** [review]
